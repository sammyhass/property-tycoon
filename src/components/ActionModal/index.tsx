import { useGameContext } from '@/hooks/useGameContext';
import { formatPrice } from '@/util/formatPrice';
import {
  Box,
  Button,
  Divider,
  Flex,
  Heading,
  Modal,
  ModalBody,
  ModalCloseButton,
  ModalContent,
  ModalHeader,
  ModalOverlay,
  Stack,
  Text,
} from '@chakra-ui/react';
import { CardAction, CardType } from '@prisma/client';
import React, { useState } from 'react';
import BoardSpace from '../Board/spaces';

// Different types of action modal we can use for our modal.
export type ActionType =
  | 'BUY' // Buying a property
  | 'MORTGAGE'
  | 'PAY_RENT' // Pay rent
  | 'ROLL' // Roll dice
  // Taking Cards
  | 'TAKE_POT_LUCK'
  | 'TAKE_OPPORTUNITY_KNOCKS'
  // ---
  | 'GO_TO_JAIL' // Player is sent to jail
  | 'FREE_PARKING' // Land on a Free Parking space
  | 'GO'; // Land on a Go space

interface ActionModalProps {
  action: ActionType | null;
  onClose: () => void;
  isOpen: boolean;
}

const actionModalComponents: Record<ActionType, React.FC<ActionModalProps>> = {
  BUY: () => <ActionModal.Buy />,
  MORTGAGE: () => <div>Mortgage</div>,
  PAY_RENT: () => <div>Pay Rent</div>,
  ROLL: props => <ActionModal.Roll {...props} />,
  GO_TO_JAIL: () => <div>Jail</div>,
  TAKE_POT_LUCK: () => <ActionModal.TakeCard cardType={'POT_LUCK'} />,
  TAKE_OPPORTUNITY_KNOCKS: () => (
    <ActionModal.TakeCard cardType="OPPORTUNITY_KNOCKS" />
  ),
  FREE_PARKING: () => <div>Free Parking</div>,
  GO: () => <div>GO!</div>,
};

export default function ActionModal(props: ActionModalProps) {
  const ActionComponent = props.action
    ? (actionModalComponents[props.action] as React.FC<ActionModalProps>)
    : null;

  return (
    <Modal onClose={props.onClose} isOpen={props.isOpen}>
      <ModalOverlay />
      <ModalContent>
        <ModalHeader>{props.action}</ModalHeader>
        <ModalCloseButton />
        <ModalBody>
          {ActionComponent && <ActionComponent {...props} />}
        </ModalBody>
      </ModalContent>
    </Modal>
  );
}

// Content for our action modal when a player is rolling the dice.
ActionModal.Roll = ({ onClose }: ActionModalProps) => {
  const { currentPlayer, move } = useGameContext();

  const [isRolling, setIsRolling] = useState(false);
  const [roll, setRoll] = useState(0);

  // Roll the dice.
  const handleRoll = () => {
    setIsRolling(true);

    if (!currentPlayer) return;
    // Choose a random number between 1 and 6
    const roll = Math.floor(Math.random() * 6) + 1;

    setTimeout(() => {
      move(currentPlayer?.token, roll);
      setRoll(roll);
      setIsRolling(false);
    }, 1000);

    if (!currentPlayer) {
      return <></>;
    }
  };
  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="4xl">ðŸŽ²</Heading>
      {isRolling && <Heading size="md">Rolling...</Heading>}
      {roll > 0 && <Heading size="sm">Dice landed on {roll}</Heading>}
      <Button onClick={() => handleRoll()}>Roll</Button>
    </Flex>
  );
};

// Content of our action modal for taking a card (pot luck or opportunity knocks)
ActionModal.TakeCard = ({ cardType }: { cardType: CardType }) => {
  const { takeCard } = useGameContext();

  const [cardAction, setCardAction] = useState<CardAction | null>(null);

  const handleTakeCard = () => {
    const action = takeCard(cardType);
    if (action) {
      setCardAction(action);
    }
  };

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="md">
        Pick a {cardType === 'POT_LUCK' ? 'Pot Luck' : 'Opportunity Knocks'}{' '}
        Card
      </Heading>
      <pre>{JSON.stringify(cardAction, null, 2)}</pre>
      <Button onClick={() => handleTakeCard()}>Take</Button>
    </Flex>
  );
};

ActionModal.Buy = () => {
  const { gameSettings, currentPlayer, state, buy, hideActionModal } =
    useGameContext();
  const [buying, setBuying] = useState(false);

  // Logic: the only way to buy a property is if it is your turn and you have landed on it.

  if (!currentPlayer) return <></>;

  const space = gameSettings?.BoardSpaces.find(
    space => space.board_position === state[currentPlayer?.token]?.pos
  );

  const property = gameSettings?.Properties.find(
    property => property.id === space?.property_id
  );

  if (!property) return <></>;

  const handleBuy = () => {
    setBuying(true);
    buy(currentPlayer?.token, property.id);
    setTimeout(() => {
      hideActionModal();
      setBuying(false);
    }, 1000);
  };

  return (
    <Box>
      <Flex direction={'column'} justify={'center'} align="center">
        <BoardSpace.Property property={property} />
      </Flex>
      <Divider my="15px" />
      <Box my="10px" p="10px" bg="#eee" borderRadius={'8px'}>
        <Heading size="md">
          {property.name} - {formatPrice(property.price ?? 0)}
        </Heading>
        <Stack>
          <Text p="0" fontWeight={'600'}>
            Rent Unimproved: {formatPrice(property.rent_unimproved ?? 0)}
            <br />
            Rent 1 House: {formatPrice(property.rent_one_house ?? 0)}
            <br />
            Rent 2 Houses: {formatPrice(property.rent_two_house ?? 0)}
            <br />
            Rent 3 Houses: {formatPrice(property.rent_three_house ?? 0)}
            <br />
            Rent 4 Houses: {formatPrice(property.rent_four_house ?? 0)}
            <br />
          </Text>
        </Stack>
      </Box>
      <Button
        colorScheme="green"
        isLoading={buying}
        onClick={() => {
          handleBuy();
        }}
      >
        Buy {property.name}
      </Button>
    </Box>
  );
};
