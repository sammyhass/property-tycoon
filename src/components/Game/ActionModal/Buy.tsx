import { useGameContext } from '@/hooks/useGameContext';
import { formatPrice } from '@/util/formatPrice';
import { Box, Button, Divider, Flex, Heading } from '@chakra-ui/react';
import React, { useState } from 'react';
import { PropertyRentInfo } from '../Board/PropertyStack';
import { BoardSpaceProperty } from '../Board/spaces';

export default function BuyContent() {
  const { gameSettings, currentPlayer, state, buy, hideActionModal } =
    useGameContext();
  const [buying, setBuying] = useState(false);

  // Logic: the only way to buy a property is if it is your turn and you have landed on it.
  if (!currentPlayer) return <></>;

  const space = gameSettings?.BoardSpaces.find(
    space => space.board_position === state[currentPlayer?.token]?.pos
  );

  const property = gameSettings?.Properties.find(
    property => property.id === space?.property_id
  );

  const group = gameSettings?.PropertyGroups.find(
    group => group.color === property?.property_group_color
  );

  if (!property) return <></>;

  const handleBuy = () => {
    setBuying(true);
    buy(currentPlayer?.token, property.id);
    setTimeout(() => {
      hideActionModal();
      setBuying(false);
    }, 1000);
  };

  return (
    <Box>
      <Flex direction={'column'} justify={'center'} align="center">
        <BoardSpaceProperty property={property} />
      </Flex>
      <Divider my="15px" />
      <Box my="10px" p="10px" bg="#eee" borderRadius={'8px'}>
        <Heading size="md">
          {property.name} - {formatPrice(property.price ?? 0)}
        </Heading>
        <PropertyRentInfo property={property} group={group} />
      </Box>
      <Flex direction="column" gap="5px">
        <Button
          w="100%"
          disabled={
            !currentPlayer ||
            !property ||
            !state[currentPlayer?.token] ||
            (state[currentPlayer?.token]?.money ?? 0) < (property?.price ?? 0) // Disable the button if you don't have enough money
          }
          colorScheme="green"
          isLoading={buying}
          onClick={() => {
            handleBuy();
          }}
        >
          {(state[currentPlayer?.token]?.money ?? 0) < (property?.price ?? 0)
            ? 'Not enough money'
            : 'Buy'}
        </Button>
        <Button w="100%" colorScheme={'blue'} onClick={() => hideActionModal()}>
          No thanks
        </Button>
      </Flex>
    </Box>
  );
}
