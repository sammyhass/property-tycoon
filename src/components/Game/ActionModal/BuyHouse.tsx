import { useGameContext } from '@/hooks/useGameContext';
import { formatPrice } from '@/util/formatPrice';
import {
  Alert,
  AlertDescription,
  AlertIcon,
  Box,
  Button,
  Flex,
  Heading,
} from '@chakra-ui/react';
import React, { useMemo } from 'react';
import { BoardSpaceProperty } from '../Board/spaces';

export default function BuyHouseContent() {
  const {
    buyHouse,
    hideBuyHouseAction,
    propertyToBuyHouseOn,
    state,
    gameSettings,
    currentPlayer,
  } = useGameContext();

  const currentHouses = useMemo(() => {
    if (!propertyToBuyHouseOn || !currentPlayer) return 0;
    return state?.[currentPlayer?.token]?.propertiesOwned?.[
      propertyToBuyHouseOn?.property_group_color
    ]?.[propertyToBuyHouseOn?.id]?.houses;
  }, [
    state,
    currentPlayer,
    propertyToBuyHouseOn,
    state[currentPlayer?.token ?? 'boot']?.propertiesOwned,
  ]);

  const houseCost = useMemo(() => {
    if (!propertyToBuyHouseOn) return 0;
    return gameSettings?.PropertyGroups.find(
      group => group.color === propertyToBuyHouseOn.property_group_color
    )?.house_cost;
  }, [propertyToBuyHouseOn, gameSettings?.PropertyGroups]);

  const hotelCost = useMemo(() => {
    if (!propertyToBuyHouseOn) return 0;
    return gameSettings?.PropertyGroups.find(
      group => group.color === propertyToBuyHouseOn.property_group_color
    )?.hotel_cost;
  }, [propertyToBuyHouseOn, gameSettings?.PropertyGroups]);

  return (
    <Box>
      <Heading size="sm">Buy a house on {propertyToBuyHouseOn?.name}?</Heading>
      <Flex w="100%" align={'center'} justify="center" my="10px">
        <BoardSpaceProperty
          property={propertyToBuyHouseOn}
          nHouses={currentHouses}
        />
      </Flex>
      {(currentHouses ?? 0) < 5 && (
        <Button
          w="100%"
          colorScheme="green"
          onClick={() => {
            if (!currentPlayer || !propertyToBuyHouseOn) return;
            buyHouse(currentPlayer?.token, propertyToBuyHouseOn.id, 1);
          }}
        >
          {`Buy a ${
            (currentHouses ?? 0) < 4 ? 'House' : 'Hotel'
          } for ${formatPrice(
            (currentHouses ?? 0) < 4 ? houseCost ?? 0 : hotelCost ?? 0
          )}`}
        </Button>
      )}

      {(currentHouses ?? 0) === 5 && (
        <Alert mt="10px">
          <AlertIcon />
          <AlertDescription>
            You have already bought a hotel on this property.
          </AlertDescription>
        </Alert>
      )}

      <Button
        onClick={hideBuyHouseAction}
        w="100%"
        colorScheme={'purple'}
        mt="10px"
      >
        All Done!
      </Button>
    </Box>
  );
}
