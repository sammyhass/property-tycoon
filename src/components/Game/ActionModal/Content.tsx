import { useGameContext } from '@/hooks/useGameContext';
import { formatPrice } from '@/util/formatPrice';
import {
  Alert,
  AlertDescription,
  Box,
  Button,
  Divider,
  Flex,
  FormControl,
  FormLabel,
  Heading,
  Radio,
  RadioGroup,
  Stack,
  Text,
} from '@chakra-ui/react';
import { CardAction, CardType } from '@prisma/client';
import React, { useCallback, useEffect, useMemo, useState } from 'react';
import { ActionModalProps } from '.';
import GameCard from '../Board/cards/Card';
import { PropertyRentInfo } from '../Board/PropertyStack';
import BoardSpace from '../Board/spaces';

// Action Modal Content Components

export const ActionModalGetOutJail = ({ action }: ActionModalProps) => {
  const {
    currentPlayer,
    rollDice,
    payBank,
    state,
    hideActionModal,
    move,
    freeFromJail,
    failedToGetOutOfJail,
  } = useGameContext();

  const [choice, setChoice] = useState<'roll' | 'pay'>('roll');

  const [roll, setRoll] = useState<[number, number] | null>(null);

  const [didRoll, setDidRoll] = useState(false);
  const [didPay, setDidPay] = useState(false);

  const handleContinue = useCallback(() => {
    if (!currentPlayer) return;
    if (didPay || (roll && roll[0] === roll[1])) {
      freeFromJail(currentPlayer.token);
      if (roll && roll[0] === roll[1]) {
        move(currentPlayer.token, roll[0] + roll[1]);
      }
    } else {
      failedToGetOutOfJail(currentPlayer.token);
    }
    hideActionModal();
  }, [
    failedToGetOutOfJail,
    currentPlayer,
    hideActionModal,
    move,
    roll,
    didPay,
    freeFromJail,
  ]);

  const handleDiceRoll = useCallback(() => {
    if (!currentPlayer) return;
    const [roll1, roll2] = rollDice();
    setRoll([roll1, roll2]);
    setDidRoll(true);
  }, [rollDice, currentPlayer]);

  const handlePayToGetOut = useCallback(() => {
    if (!currentPlayer) return;
    payBank(currentPlayer?.token, 50);
    setDidPay(true);
  }, [currentPlayer, payBank]);

  if (!currentPlayer) return <></>;
  return (
    <Flex direction={'column'} gap="5px">
      <Heading size="md" py="10px">
        You&apos;sre in Jail!
      </Heading>

      {!(didRoll || didPay) && (
        <FormControl mb="20px">
          <FormLabel>
            <Text fontSize="sm">
              Try to get out of jail by rolling doubles (
              {3 - (state[currentPlayer.token]?.turnsInJail ?? 0)} tries left)
              or paying the fine of $50.
            </Text>
          </FormLabel>
          <RadioGroup
            value={choice}
            onChange={v =>
              v === 'roll' ? setChoice(v) : v === 'pay' ? setChoice(v) : null
            }
            defaultValue="roll"
          >
            <Stack direction="column" spacing="5px">
              <Radio value="roll">
                <Text>ðŸŽ² Roll Dice</Text>
              </Radio>
              <Radio value="pay">
                <Text>ðŸ’¸ Pay $50 to get out</Text>
              </Radio>
            </Stack>
          </RadioGroup>
        </FormControl>
      )}

      {choice === 'roll' ? (
        <Box>
          <Heading size="sm">
            Try to get out of jail by rolling doubles (
            {3 - (state[currentPlayer.token]?.turnsInJail ?? 0)} tries left)
          </Heading>
          <Box p="10px" bg="#eee" borderRadius={'8px'}>
            <Heading textAlign={'center'}>ðŸŽ² ðŸŽ²</Heading>
            {roll ? (
              <Box>
                <Text textAlign={'center'}>You Rolled</Text>
                <Stack
                  direction="row"
                  spacing={'5px'}
                  justify="center"
                  fontSize={'lg'}
                >
                  <Text fontWeight={'bold'}>{roll[0]}</Text>
                  <Text>and</Text>
                  <Text fontWeight={'bold'}>{roll[1]}</Text>
                </Stack>
                {roll[0] === roll[1] && (
                  <Alert status="success">
                    <AlertDescription>
                      You rolled doubles! You can get out of jail!
                    </AlertDescription>
                  </Alert>
                )}
              </Box>
            ) : (
              <Button onClick={handleDiceRoll} w="100%" colorScheme={'red'}>
                Roll Dice
              </Button>
            )}
          </Box>
        </Box>
      ) : (
        <Box>
          <Heading size="sm">Pay Â£50 to get out of jail now?</Heading>
          <Button
            colorScheme={'purple'}
            onClick={handlePayToGetOut}
            w="100%"
            isDisabled={
              !currentPlayer || (state[currentPlayer.token]?.money ?? 0) < 50
            }
          >
            Pay Â£50 to get out
          </Button>
        </Box>
      )}
      {(didRoll || didPay) && (
        <Button onClick={handleContinue}>
          {roll
            ? roll[0] === roll[1] || didPay
              ? 'Get out Now'
              : 'Roll again next turn.'
            : 'Ok'}
        </Button>
      )}
    </Flex>
  );
};

// Content for our action modal when a player is rolling the dice.
export const ActionModalRoll = () => {
  const { currentPlayer, move, rollDice } = useGameContext();

  const [isRolling, setIsRolling] = useState(false);
  const [isMoving, setIsMoving] = useState(false);

  const [diceRoll, setDiceRoll] = useState<[number, number] | null>(null);

  const handleMove = useCallback(() => {
    if (!currentPlayer) return;
    setIsMoving(true);
    setTimeout(() => {
      move(currentPlayer.token, diceRoll![0] + diceRoll![1]);
      setIsMoving(false);
    }, 500);
  }, [currentPlayer, diceRoll, move]);

  // Roll the dice.
  const handleRoll = useCallback(() => {
    setIsRolling(true);

    if (!currentPlayer) return;
    // Choose a random number between 1 and 12
    setTimeout(() => {
      const [dice1, dice2] = rollDice();
      setDiceRoll([dice1, dice2]);
      setIsRolling(false);
    }, 500);
  }, [currentPlayer, rollDice]);

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="4xl">ðŸŽ² ðŸŽ²</Heading>
      {isRolling && <Heading size="md">Rolling...</Heading>}
      {diceRoll && (
        <Stack spacing={4} justify="center" align={'center'}>
          <Text>You rolled:</Text>
          <Flex justify={'center'} gap="10px" align="center">
            <Heading size="md">{diceRoll[0]}</Heading>
            <Text>{` and `}</Text>
            <Heading size="md">{diceRoll[1]}</Heading>
          </Flex>
        </Stack>
      )}
      <Button
        w="100%"
        colorScheme={diceRoll ? 'green' : 'purple'}
        onClick={diceRoll ? handleMove : handleRoll}
        mt="10px"
        isLoading={isRolling || isMoving}
        disabled={isRolling || isMoving}
      >
        {diceRoll ? 'Move' : 'Roll'}
      </Button>
    </Flex>
  );
};

export const ActionModalGo = ({}) => {
  const { payBank, currentPlayer, hideActionModal } = useGameContext();

  const [loading, setLoading] = useState(false);

  const collect = useCallback(() => {
    setLoading(true);
    if (!currentPlayer) return;
    payBank(currentPlayer.token, -200);

    setTimeout(() => {
      setLoading(false);

      hideActionModal();
    }, 500);
  }, [currentPlayer, payBank, hideActionModal]);

  if (!currentPlayer) return <></>;

  return (
    <Flex direction={'column'} gap="5px" align={'center'}>
      <Heading size="4xl">ðŸŽ‰</Heading>
      <Heading size="md">You landed on Go!</Heading>
      <Box>Collect {formatPrice(200)} from the bank!</Box>
      <Button
        w="100%"
        colorScheme={'green'}
        isLoading={loading}
        onClick={collect}
      >
        Collect Â£200
      </Button>
    </Flex>
  );
};

// Content of our action modal for taking a card (pot luck or opportunity knocks)
export const ActionModalTakeCard = ({ cardType }: { cardType: CardType }) => {
  const {
    takeCard,
    payBank,
    currentPlayer,
    gameSettings,
    goto,
    sendToJail,
    hideActionModal,
  } = useGameContext();

  const [cardAction, setCardAction] = useState<CardAction | null>(null);

  const handleTakeCard = () => {
    const action = takeCard(cardType);
    if (action) {
      setCardAction(action);
    }
  };

  const performAction = useCallback(() => {
    if (!cardAction || !currentPlayer?.token) return;

    hideActionModal();

    switch (cardAction.action_type) {
      case 'EARN_FROM_BANK':
        payBank(currentPlayer?.token, -(cardAction?.cost ?? 0));
        break;
      case 'PAY_BANK':
        payBank(currentPlayer?.token, cardAction?.cost ?? 0);
        break;
      case 'GO_TO_JAIL':
        sendToJail(currentPlayer?.token);
        break;
      case 'GO_TO_PROPERTY':
        const property = gameSettings?.BoardSpaces?.find(
          space => space.property_id === cardAction?.property_id
        );
        if (property) {
          goto(currentPlayer?.token, property.board_position);
        }
      default:
        break;
    }
  }, [
    cardAction,
    currentPlayer,
    gameSettings,
    hideActionModal,
    payBank,
    sendToJail,
    goto,
  ]);

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="md">
        Pick a {cardType === 'POT_LUCK' ? 'Pot Luck' : 'Opportunity Knocks'}{' '}
        Card
      </Heading>
      {cardAction && (
        <GameCard
          {...cardAction}
          propertyName={
            gameSettings?.Properties?.find(
              property => property.id === cardAction?.property_id
            )?.name ?? undefined
          }
        />
      )}
      <Button
        colorScheme={'green'}
        mt={'10px'}
        w="100%"
        onClick={() => (cardAction ? performAction() : handleTakeCard())}
      >
        {cardAction ? 'Perform Card Action' : 'Take'}
      </Button>
    </Flex>
  );
};

export const ActionModalBuy = () => {
  const { gameSettings, currentPlayer, state, buy, hideActionModal } =
    useGameContext();
  const [buying, setBuying] = useState(false);

  // Logic: the only way to buy a property is if it is your turn and you have landed on it.

  if (!currentPlayer) return <></>;

  const space = gameSettings?.BoardSpaces.find(
    space => space.board_position === state[currentPlayer?.token]?.pos
  );

  const property = gameSettings?.Properties.find(
    property => property.id === space?.property_id
  );

  if (!property) return <></>;

  const handleBuy = () => {
    setBuying(true);
    buy(currentPlayer?.token, property.id);
    setTimeout(() => {
      hideActionModal();
      setBuying(false);
    }, 1000);
  };

  return (
    <Box>
      <Flex direction={'column'} justify={'center'} align="center">
        <BoardSpace.Property property={property} />
      </Flex>
      <Divider my="15px" />
      <Box my="10px" p="10px" bg="#eee" borderRadius={'8px'}>
        <Heading size="md">
          {property.name} - {formatPrice(property.price ?? 0)}
        </Heading>
        <PropertyRentInfo property={property} />
      </Box>
      <Flex direction="column" gap="5px">
        <Button
          w="100%"
          disabled={
            !currentPlayer ||
            !property ||
            !state[currentPlayer?.token] ||
            (state[currentPlayer?.token]?.money ?? 0) < (property?.price ?? 0) // Disable the button if you don't have enough money
          }
          colorScheme="green"
          isLoading={buying}
          onClick={() => {
            handleBuy();
          }}
        >
          {(state[currentPlayer?.token]?.money ?? 0) < (property?.price ?? 0)
            ? 'Not enough money'
            : 'Buy'}
        </Button>
        <Button w="100%" colorScheme={'blue'} onClick={() => hideActionModal()}>
          No thanks
        </Button>
      </Flex>
    </Box>
  );
};

export const ActionModalTax = () => {
  const { gameSettings, state, currentPlayer, payBank, hideActionModal } =
    useGameContext();

  if (!gameSettings || !currentPlayer) return <></>;

  const space = gameSettings?.BoardSpaces.find(
    space => space.board_position === (state[currentPlayer?.token]?.pos ?? 0)
  );

  const taxAmount = space?.tax_cost ?? 0;

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="md">Pay the Tax</Heading>
      <Text fontSize={'4xl'}>{formatPrice(taxAmount)}</Text>
      <Button
        w={'100%'}
        colorScheme={'red'}
        onClick={() => {
          payBank(currentPlayer?.token, taxAmount);
          setTimeout(() => {
            hideActionModal();
          }, 1000);
        }}
      >
        Pay
      </Button>
    </Flex>
  );
};

export const ActionModalPayRent = () => {
  const {
    payPlayer,
    gameSettings,
    currentPlayer,
    state,
    isOwned,
    calculateRent,
    hideActionModal,
  } = useGameContext();

  const space = currentPlayer?.token
    ? gameSettings?.BoardSpaces.find(
        space =>
          space.board_position === (state[currentPlayer?.token]?.pos ?? 0)
      )
    : null;

  const property = gameSettings?.Properties.find(
    property => property.id === space?.property_id
  );

  const owner = isOwned(property?.id ?? '');

  const rent = useMemo(() => {
    if (!property) return 0;
    return calculateRent(property?.id);
  }, [calculateRent, property]);

  if (!owner || !property) return <></>;

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="md">Pay the Rent</Heading>
      {property && <BoardSpace.Property property={property} />}
      <Text fontSize={'4xl'}>{formatPrice(rent ?? 0)}</Text>
      <Button
        w={'100%'}
        colorScheme={'red'}
        onClick={() => {
          if (!currentPlayer) return;
          payPlayer(currentPlayer?.token, owner?.token, rent);

          setTimeout(() => {
            hideActionModal();
          }, 1000);
        }}
      >
        Pay
      </Button>
    </Flex>
  );
};

export const ActionModalFreeParking = () => {
  const {
    landedOnFreeParking,
    currentPlayer,
    hideActionModal,
    totalOnFreeParking,
  } = useGameContext();

  useEffect(() => {
    if (!currentPlayer) return;
    const v = landedOnFreeParking(currentPlayer.token);
  }, [currentPlayer, landedOnFreeParking]);

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="md">Free Parking</Heading>

      <Text>
        The total amount on free parking is {formatPrice(totalOnFreeParking)},
        It&apos;s all yours now
      </Text>
      <Button
        w={'100%'}
        colorScheme={'blue'}
        onClick={() => {
          hideActionModal();
        }}
      >
        Continue
      </Button>
    </Flex>
  );
};

export const ActionModalGoToJail = () => {
  const { sendToJail, currentPlayer, hideActionModal, gameSettings } =
    useGameContext();

  const handleGoToJail = () => {
    if (!currentPlayer) return;
    sendToJail(currentPlayer?.token);

    setTimeout(() => {
      hideActionModal();
    }, 1000);
  };

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="md">Go To Jail</Heading>
      <Text>You have been sent to jail</Text>
      <Button onClick={() => handleGoToJail()}>Go To Jail</Button>
    </Flex>
  );
};
