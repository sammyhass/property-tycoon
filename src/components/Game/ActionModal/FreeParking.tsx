import { useGameContext } from '@/hooks/useGameContext';
import { formatPrice } from '@/util/formatPrice';
import { Button, Flex, Heading, Text } from '@chakra-ui/react';
import React, { useEffect } from 'react';

export default function ActionModalFreeParking() {
  const {
    landedOnFreeParking,
    currentPlayer,
    hideActionModal,
    totalOnFreeParking,
  } = useGameContext();

  useEffect(() => {
    if (!currentPlayer) return;
    const v = landedOnFreeParking(currentPlayer.token);
  }, [currentPlayer, landedOnFreeParking]);

  return (
    <Flex direction={'column'} justify={'center'} align="center">
      <Heading size="md">Free Parking</Heading>

      <Text>
        The total amount on free parking is {formatPrice(totalOnFreeParking)},
        It&apos;s all yours now
      </Text>
      <Button
        w={'100%'}
        colorScheme={'blue'}
        onClick={() => {
          hideActionModal();
        }}
      >
        Continue
      </Button>
    </Flex>
  );
}
