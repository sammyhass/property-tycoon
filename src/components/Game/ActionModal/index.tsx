import { Box, Heading } from '@chakra-ui/react';
import React, { useMemo } from 'react';
import HUD from '../HUD';
import {
  ActionModalBuy,
  ActionModalFreeParking,
  ActionModalGetOutJail,
  ActionModalGo,
  ActionModalGoToJail,
  ActionModalPayRent,
  ActionModalRoll,
  ActionModalTakeCard,
  ActionModalTax,
} from './Content';

// Different types of action modal we can use for our modal.
export type ActionType =
  | 'BUY' // Buying a property
  | 'PAY_RENT' // Pay rent
  | 'ROLL' // Roll dice
  // Taking Cards
  | 'TAKE_POT_LUCK'
  | 'TAKE_OPPORTUNITY_KNOCKS'
  // ---
  | 'GO_TO_JAIL' // Player is sent to jail
  | 'GET_OUT_OF_JAIL' // Player attempts to get out of jail (shows instead)
  | 'FREE_PARKING' // Land on a Free Parking space
  | 'TAX' // Pay tax
  // TODO:
  | 'MORTGAGE'
  | 'GO'; // Land on a Go space

export interface ActionModalProps {
  action: ActionType | null;
}

const actionModalComponents: Record<ActionType, React.FC<ActionModalProps>> = {
  BUY: ActionModalBuy,
  MORTGAGE: () => <div>Mortgage</div>,
  PAY_RENT: ActionModalPayRent,
  GET_OUT_OF_JAIL: ActionModalGetOutJail,
  ROLL: ActionModalRoll,
  GO_TO_JAIL: ActionModalGoToJail,
  TAKE_POT_LUCK: () => <ActionModalTakeCard cardType={'POT_LUCK'} />,
  TAKE_OPPORTUNITY_KNOCKS: () => (
    <ActionModalTakeCard cardType="OPPORTUNITY_KNOCKS" />
  ),
  FREE_PARKING: ActionModalFreeParking,
  GO: ActionModalGo,
  TAX: ActionModalTax,
};

export default function ActionSidebar(props: ActionModalProps) {
  const ActionComponent = useMemo(
    () =>
      props.action
        ? (actionModalComponents[props.action] as React.FC<ActionModalProps>)
        : null,
    [props.action]
  );

  return (
    <Box minW={'400px'} bg="white" px="20px" h="100vh">
      <Heading>
        {props.action
          ?.split('_')
          .map(w => w[0].toUpperCase() + w.slice(1).toLowerCase())
          .join(' ')}
      </Heading>
      <Box minH="70%">{ActionComponent && <ActionComponent {...props} />}</Box>
      <HUD />
    </Box>
  );
}
