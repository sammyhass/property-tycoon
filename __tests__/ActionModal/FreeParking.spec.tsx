import { ActionModalFreeParking } from '@/components/Game/ActionModal/Content';
import { PlayerState, TokenType } from '@/hooks/useGameContext';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { fakePlayer, renderWithGameContext } from '../fakers';

describe('Free Parking Action Modal', () => {
  it('should match snapshot', () => {
    const { container } = renderWithGameContext(<ActionModalFreeParking />, {
      currentPlayer: fakePlayer('boot'),
      state: {
        boot: {
          inJail: false,
          money: 400,
          pos: 1,
          propertiesOwned: [],
        },
        cat: {
          inJail: false,
          money: 1200,
          pos: 1,
          propertiesOwned: [],
        },
      },
      totalOnFreeParking: 100,
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correct text', () => {
    const { container } = renderWithGameContext(<ActionModalFreeParking />, {
      currentPlayer: fakePlayer('boot'),
      state: {
        boot: {
          inJail: false,
          money: 400,
          pos: 1,
          propertiesOwned: [],
        },
        cat: {
          inJail: false,
          money: 1200,
          pos: 1,
          propertiesOwned: [],
        },
      },
      totalOnFreeParking: 100,
    });

    expect(container.textContent).toContain('Free Parking');
  });

  it('should render correct amount of money', () => {
    const { container } = renderWithGameContext(<ActionModalFreeParking />, {
      currentPlayer: fakePlayer('boot'),
      state: {
        boot: {
          inJail: false,
          money: 400,
          pos: 1,
          propertiesOwned: [],
        },
        cat: {
          inJail: false,
          money: 1200,
          pos: 1,
          propertiesOwned: [],
        },
      },
      totalOnFreeParking: 100,
    });

    expect(container.textContent).toContain('Â£100');
  });

  it('on clicking collect button, the total in free parking is collected by the current player', async () => {
    let totalOnFreeParking = 200;
    let state: PlayerState = {
      boot: {
        inJail: false,
        money: 400,
        pos: 1,
        propertiesOwned: [],
      },
      cat: {
        inJail: false,
        money: 1200,
        pos: 1,
        propertiesOwned: [],
      },
    };
    const landedOnFreeParking = jest.fn().mockImplementation((p: TokenType) => {
      if (!state[p]) return;
      state = {
        ...state,
        [p]: {
          ...state[p],
          money: (state[p]?.money ?? 0) + totalOnFreeParking,
        },
      };

      totalOnFreeParking = 0;
    });

    const { container } = renderWithGameContext(<ActionModalFreeParking />, {
      currentPlayer: fakePlayer('boot'),
      state,
      totalOnFreeParking,
      landedOnFreeParking,
    });

    const button = container.querySelector('button');

    await act(async () => {
      button?.click();
      await new Promise(resolve => setTimeout(resolve, 1000));
    });

    expect(totalOnFreeParking).toBe(0);
    // Check that the money is added to the current player
    expect(state.boot?.money).toBe(600);
    // And not to the other player
    expect(state.cat?.money).toBe(1200);
  });
});
