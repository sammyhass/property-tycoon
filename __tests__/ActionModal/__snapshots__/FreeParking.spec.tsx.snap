// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Free Parking Action Modal should match snapshot 1`] = `
<div
  class="css-u6v8er"
>
  <h2
    class="chakra-heading css-0"
  >
    Free Parking
  </h2>
  <p
    class="chakra-text css-0"
  >
    The total amount on free parking is 
    Â£100
    , It's all yours now
  </p>
  <button
    class="chakra-button css-uxt1e8"
    type="button"
  >
    Continue
  </button>
</div>
`;
